compiler/rustc_parse/src/parser/diagnostics.rs

```diff
@@ -1170,16 +1171,20 @@ impl<'a> Parser<'a> {
         impl_dyn_multi: bool,
         ty: &Ty,
     ) {
+        #[derive(SessionDiagnostic)]
+        #[error(slug = "parser-maybe-report-ambiguous-plus")]
+        struct AmbiguousPlus {
+            pub sum_with_parens: String,
+            #[primary_span]
+            #[suggestion(code = "{sum_with_parens}")]
+            pub span: Span,
+        }
+
         if matches!(allow_plus, AllowPlus::No) && impl_dyn_multi {
-            let sum_with_parens = format!("({})", pprust::ty_to_string(&ty));
-            self.struct_span_err(ty.span, "ambiguous `+` in a type")
-                .span_suggestion(
-                    ty.span,
-                    "use parentheses to disambiguate",
-                    sum_with_parens,
-                    Applicability::MachineApplicable,
-                )
-                .emit();
+            self.sess.emit_err(AmbiguousPlus {
+                sum_with_parens: format!("({})", pprust::ty_to_string(&ty)),
```

Comment:
Could we avoid a `format!` call here by just adding the parens to the `code = "({sum_with_parens})"` of the attribute?